### PR TITLE
hook up runtime intrinsics

### PR DESCRIPTION
**Pull request by vtjnash**
This long-overdue PR makes the intrinsic functions work just like normal builtin functions instead of requiring special rules in the compiler.

incorporates parts of jn/jlinterpreter and #14918

### CODE REVIEW

File: src/gf.c [hunk 1]
```diff
@@ -154,10 +154,14 @@ JL_DLLEXPORT jl_value_t *jl_methtable_lookup(jl_methtable_t *mt, jl_tupletype_t
 
 JL_DLLEXPORT jl_method_t *jl_new_method_uninit(void);
 static jl_function_t *jl_new_generic_function_with_supertype(jl_sym_t *name, jl_module_t *module, jl_datatype_t *st, int iskw);
-jl_value_t *jl_mk_builtin_func(const char *name, jl_fptr_t fptr)
+void jl_mk_builtin_func(jl_datatype_t *dt, const char *name, jl_fptr_t fptr)
 {
     jl_sym_t *sname = jl_symbol(name);
-    jl_value_t *f = jl_new_generic_function_with_supertype(sname, jl_core_module, jl_builtin_type, 0);
+    if (dt == NULL) {
+       jl_value_t *f = jl_new_generic_function_with_supertype(sname, jl_core_module, jl_builtin_type, 0);
```
Comment (tkelman) [inline src/gf.c]: indent slightly off here

File: src/intrinsics.h [hunk 2]
```diff
@@ -152,21 +153,25 @@ extern "C"
 JL_CALLABLE(jl_f_intrinsic_call)
 {
     JL_NARGSV(intrinsic_call, 1);
-    JL_TYPECHK(intrinsic_call, intrinsic, args[0]);
-    enum intrinsic f = (enum intrinsic)*(uint32_t*)jl_data_ptr(args[0]);
+    JL_TYPECHK(intrinsic_call, intrinsic, F);
```
Comment (tkelman) [inline src/intrinsics.h]: where does capital `F` come from?

Comment (yuyichao) [inline src/intrinsics.h]: `JL_CALLABLE`. `args` comes from the same place. I personally agree it's confusing and one way to make it more explicit is to pass in the variable names a macro parameters, i.e. `JL_CALLABLE(..., F, args)`.

Anyway, unrelated to this PR.

Comment (JeffBezanson) [inline src/intrinsics.h]: Ideally we will stop using `JL_CALLABLE`, and instead use function-specific signatures that are more efficient where possible.

**Pull request merged**